### PR TITLE
feat: add cosmos scheduler

### PR DESCRIPTION
**Pull request by giacope**
**What does this PR do, and why do we need it?**

This PR adds the Cosmos Scheduler, a new component that will create and delete Cosmos CRs based on the creation and deletion of NFS Ingestion Buckets.

**Special notes for your reviewers**:

Vendor dependencies will come in another PR to simplify the review process.

### CODE REVIEW

File: cosmos/scheduler/Dockerfile
Comment (giacope) [inline cosmos/scheduler/Dockerfile]: Probably need a better Dockerfile (that doesn't depend on a locally built binary)

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go:125]: Not sure if this the best way to tackle the scheduler logic. Open to suggestions...

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go:190]: Be more specific about location types instead of relying on `strstr`.

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go:222]: Opinions for a way for tackling this notification system are welcome :)

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go]: Need a better method name

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go]: This could be wrapped in another method

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go]: Downto here

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [inline cosmos/scheduler/cmd/scheduler.go:237]: Be more generic about the location for future use cases

File: cosmos/scheduler/main.go
Comment (giacope) [inline cosmos/scheduler/main.go:82]: Suggestions for a better way of dealing with the configuration of the scheduler are welcome :)

File: cosmos/scheduler/pkg/pensieve.go
Comment (giacope) [inline cosmos/scheduler/pkg/pensieve.go:112]: change for ==

File: cosmos/scheduler/pkg/pensieve.go
Comment (giacope) [inline cosmos/scheduler/pkg/pensieve.go:221]: Make this function error out properly when the endpoint is malformed.

File: kubernetes/zenko/templates/cosmos-scheduler/deployment.yaml
Comment (giacope) [inline kubernetes/zenko/templates/cosmos-scheduler/deployment.yaml:2]: Add proper templates and meta stuff to this deployment.

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: it might be safer to set `batchSize` for the watch to only return one document at a time.

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: for clarification, why does the `handler` need to quit for each "overlay update"? I may be missing something, but it doesn't look like the `handler` does anything with that event. For each `__metadata` changes, a check for the latest overlay configuration is always performed, so the `handler` routine doesn't need to "restart"

If that quit doesn't matter, I believe the `handler` can be run without the loop?

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: I think `waitForOverlayUpdate` is creating a new `changeStream` for each  it's called.
maybe move the loop logic into the method?

an example using `sync.WaitGroup`:
(not sure how this should exit gracefully)

```
func (s *Scheduler) Run() {
          ...
          wg := sync.WaitGroup
          defer wg.Wait()
          wg.Add(1)
          ... handler(msgCh, &wg) goroutine
          wg.Add(1)
          go waitForOverlayUpdate(msgCh, &wg)
          ...
}

func (s *Schedule) waitForOverlayUpdate(ch chan<- string, wg *sync.WaitGroup) {
        defer wg.Done()
        collection := s.MongodbClient.Database("metadata").Collection("PENSIEVE")
	ctx := context.Background()
	cur, err := collection.Watch(ctx, mongo.Pipeline{})
        defer cur.Close(ctx)
	if err != nil {
		log.Println(err)
	}
        for {
                if cur.Next(ctx) {
                        ch <- "update overlay"
                } else {
                        if err := cur.Err(); err != nil {
                                log.Println(err)
                        }
                }
        }
}
```

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [reply cosmos/scheduler/cmd/scheduler.go]: The reason for that is because every time the handler restarts, it starts a new `s.watchBucketEvents()` which will create a new watch stream with a new set of buckets.

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: from our discussion, the idea of making this section a goroutine and closing the cursor on quit does seem to work. on cursor close, if `cursor.Next(ctx)` is block for a new entry, `false` will be returned and it will unblock. the  `cursor.Err()` will also be set to `ended session was used`.

File: cosmos/scheduler/cmd/scheduler.go
Comment (ssalaues) [inline cosmos/scheduler/cmd/scheduler.go:171]: Add a todo here as it would conflict if there are multiple Zenkos within one namespace.

**Review by ssalaues — APPROVED** [review]

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [reply cosmos/scheduler/cmd/scheduler.go:171]: Done

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: should this be in the `else` statement? correct me if I am wrong, this case doesn't seem reachable if `cur.Next() == false`

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: the function doesn't end for the defer to kick in

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go:142]: should `return` here?

File: cosmos/scheduler/cmd/scheduler.go
Comment (alexanderchan-scality) [inline cosmos/scheduler/cmd/scheduler.go]: should check for `overlayUpdates != nil`?

**Review by alexanderchan-scality — COMMENTED** [review]
first pass comments

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [reply cosmos/scheduler/cmd/scheduler.go]: That is correct

File: cosmos/scheduler/cmd/scheduler.go
Comment (giacope) [reply cosmos/scheduler/cmd/scheduler.go:142]: no, removing the `defer` instead...

File: cosmos/README.md
Comment (alexanderchan-scality) [inline cosmos/README.md]: minor
```suggestion
This directory holds most the code that corresponds to the Cosmos ingestion framework.
```

File: cosmos/README.md
Comment (alexanderchan-scality) [inline cosmos/README.md]: minor
```suggestion
The subdirectory layout is as follows:
```

File: cosmos/README.md
Comment (alexanderchan-scality) [inline cosmos/README.md]: minor
```suggestion
    This directory holds the code for the Orchestration Daemon.
```

File: cosmos/README.md
Comment (alexanderchan-scality) [inline cosmos/README.md]: minor
```suggestion
    This directory contains the dependencies for all the Cosmos Go packages.
```

**Review by alexanderchan-scality — APPROVED** [review]

**Comment by giacope** [comment]
@bert-e approve

**Review by tmacro — APPROVED** [review]

**Pull request merged**
